ext/xsl/xsltprocessor.c

```diff
@@ -21,75 +21,29 @@
 
 #include "php.h"
 #include "php_xsl.h"
+#include <libxslt/variables.h>
 #include "ext/libxml/php_libxml.h"
 
-/* {{{ php_xsl_xslt_string_to_xpathexpr()
-   Translates a string to a XPath Expression */
-static char *php_xsl_xslt_string_to_xpathexpr(const char *str)
-{
-	const xmlChar *string = (const xmlChar *)str;
-
-	xmlChar *value;
-	int str_len;
 
-	str_len = xmlStrlen(string) + 3;
-
-	if (xmlStrchr(string, '"')) {
-		if (xmlStrchr(string, '\'')) {
-			php_error_docref(NULL, E_WARNING, "Cannot create XPath expression (string contains both quote and double-quotes)");
-			return NULL;
-		}
-		value = (xmlChar*) safe_emalloc (str_len, sizeof(xmlChar), 0);
-		snprintf((char*)value, str_len, "'%s'", string);
-	} else {
-		value = (xmlChar*) safe_emalloc (str_len, sizeof(xmlChar), 0);
-		snprintf((char *)value, str_len, "\"%s\"", string);
-	}
-	return (char *) value;
-}
-/* }}} */
-
-/* {{{ php_xsl_xslt_make_params()
-   Translates a PHP array to a libxslt parameters array */
-static char **php_xsl_xslt_make_params(HashTable *parht, int xpath_params)
+static zend_result php_xsl_xslt_apply_params(xsltTransformContextPtr ctxt, HashTable *params)
 {
-
-	int parsize;
-	zval *value;
-	char *xpath_expr;
 	zend_string *string_key;
-	char **params = NULL;
-	int i = 0;
-
-	parsize = (2 * zend_hash_num_elements(parht) + 1) * sizeof(char *);
-	params = (char **)safe_emalloc((2 * zend_hash_num_elements(parht) + 1), sizeof(char *), 0);
-	memset((char *)params, 0, parsize);
+	zval *value;
 
-	ZEND_HASH_MAP_FOREACH_STR_KEY_VAL(parht, string_key, value) {
+	ZEND_HASH_MAP_FOREACH_STR_KEY_VAL(params, string_key, value) {
 		ZEND_ASSERT(string_key != NULL);
-		if (Z_TYPE_P(value) != IS_STRING) {
-			if (!try_convert_to_string(value)) {
-				efree(params);
-				return NULL;
-			}
-		}
+		/* Already a string because of setParameter() */
+		ZEND_ASSERT(Z_TYPE_P(value) == IS_STRING);
 
-		if (!xpath_params) {
-			xpath_expr = php_xsl_xslt_string_to_xpathexpr(Z_STRVAL_P(value));
-		} else {
-			xpath_expr = estrndup(Z_STRVAL_P(value), Z_STRLEN_P(value));
-		}
-		if (xpath_expr) {
-			params[i++] = estrndup(ZSTR_VAL(string_key), ZSTR_LEN(string_key));
-			params[i++] = xpath_expr;
+		int result = xsltQuoteOneUserParam(ctxt, (const xmlChar *) ZSTR_VAL(string_key), (const xmlChar *) Z_STRVAL_P(value));
```

Comment:
Won't this potentially cause issues with values (or even keys...) that have a nul byte? Seems with the manual quoting this would have been taken care of?

Maybe this needs to be checked in ``setParameter()`` that the provided strings do not contain nul bytes?

Reply:
Yes this causes issues with NUL bytes.

> Seems with the manual quoting this would have been taken care of?

Nope, even today if you do this:
```php
$xslt->setParameter('', 'foo', "a\0b");
var_dump($xslt->getParameter('', 'foo'));
```

you get the output `string(3) "ab"`, but only the letter a is outputted.

> Maybe this needs to be checked in setParameter() that the provided strings do not contain nul bytes?

That would be best.

Reply:
Ok implemented the null byte check now